File: src/main/java/stormpot/fuzzer/Fuzzer.java
package stormpot.fuzzer;

import java.util.ArrayList;
import java.util.Deque;
import java.util.List;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.TimeUnit;

import stormpot.Allocator;
import stormpot.Config;
import stormpot.Expiration;
import stormpot.LifecycledPool;
import stormpot.Pool;
import stormpot.PoolException;
import stormpot.Timeout;

public class Fuzzer {
  private static final Timeout DEFAULT_TIMEOUT =
      new Timeout(1, TimeUnit.SECONDS);
  
  private static final ClaimStrategy DEFAULT_CLAIM = new ClaimStrategy() {
    @Override
    public void invoke(Pool<GenericPoolable> pool)
        throws PoolException, InterruptedException {
      pool.claim(DEFAULT_TIMEOUT).release();
    }
  };
  
  private ThreadFactory threadFactory;
  private int workerThreadCount;
  private List<ThreadBuilder> serviceThreads;
  private Deque<ClaimStrategy> claimStrategies;
  private PoolFactory poolFactory;
  private Config<GenericPoolable> config;
  
  public Fuzzer() {
    threadFactory = null;
    serviceThreads = new ArrayList<>();
    config = new Config<GenericPoolable>();
  }
  
  public void setWorkerThreadCount(int count) {
    workerThreadCount = count;
  }
  
  public void addServiceThread(ThreadBuilder serviceThread) {
    serviceThreads.add(serviceThread);
  }
  
  public void addClaimStrategy(ClaimStrategy strategy) {
    claimStrategies.add(strategy);
  }
  
  public void setExpiration(Expiration<GenericPoolable> expiration) {
    config.setExpiration(expiration);
  }
  
  public void setPoolFixture(PoolFactory factory) {
    poolFactory = factory;
  }
  
  public void setAllocator(Allocator<GenericPoolable> allocator) {
    config.setAllocator(allocator);
  }
  
  public void setPoolSize(int size) {
    config.setSize(size);
  }
  
  public void run(long runTimeMillis) throws InterruptedException {
    Pool<GenericPoolable> pool = poolFactory.createPool(config);
    
    List<Thread> threads = new ArrayList<>();
    addAllServiceThreads(pool, threads);
    addAllWorkerThreads(pool, threads);
    
    startAll(threads);
    Thread.sleep(runTimeMillis);
    interruptAll(threads);
    shutDown(pool);
    joinAll(threads);
  }

  private void startAll(List<Thread> threads) {
    for (Thread thread : threads) {
      thread.start();
    }
  }

  private void interruptAll(List<Thread> threads) {
    for (Thread thread : threads) {
      thread.interrupt();
    }
  }

  private void shutDown(Pool<?> pool) throws InterruptedException {
    if (pool instanceof LifecycledPool) {
      ((LifecycledPool<?>) pool).shutdown().await(DEFAULT_TIMEOUT);
    }
  }
  
  private void joinAll(List<Thread> threads) throws InterruptedException {
    for (Thread thread : threads) {
      thread.join();
    }
  }

  private void addAllServiceThreads(
      Pool<GenericPoolable> pool,
      List<Thread> threads) {
    for (ThreadBuilder builder : serviceThreads) {
      threads.add(builder.buildThread(pool));
    }
  }

  private void addAllWorkerThreads(
      Pool<GenericPoolable> pool,
      List<Thread> threads) {
    for (int i = 0; i < workerThreadCount; i++) {
      ClaimStrategy strategy = coalesce(claimStrategies.poll(), DEFAULT_CLAIM);
      threads.add(createWorkerThread(strategy, pool));
    }
    
    if (!claimStrategies.isEmpty()) {
      String msg = "" +
      		"WARNING: Not enough worker threads for all ClaimStrategies." +
      		" %s strategies left!\n";
      System.err.printf(msg, claimStrategies.size());
    }
  }

  private <T> T coalesce(T a, T b) {
    return a == null? b : a;
  }

  private Thread createWorkerThread(
      final ClaimStrategy strategy,
      final Pool<GenericPoolable> pool) {
    Runnable runnable = new Runnable() {
      @Override
      public void run() {
        try {
          for (;;) {
            strategy.invoke(pool);
          }
        } catch (PoolException e) {
          e.printStackTrace();
          // TODO we might get this because the pool is shut down.
          // Figure out a nice way to deal with that.
          // Ideally, the fuzzer should run such that we can trust, that any
          // stack trace produced, means that we've encountered a bug.
        } catch (InterruptedException e) {
          // stop signal.
        }
      }
    };
    return threadFactory.newThread(runnable);
  }
}

File: src/main/scala/stormpot/fuzzer/Fuzzer.scala
package stormpot.fuzzer

import stormpot._
import java.util.concurrent.{CountDownLatch, TimeUnit}
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.LockSupport
import collection.Iterator
import scala.collection.mutable.ArrayStack

class Fuzzer(
    val config: Config[Poolable],
    var poolFactory: PoolFactory = null,
    var absoluteThreadCount: Int = -1,
    var relativeThreadFactor: Double = -1.0,
    var poolSize: Int = 10,
    var fuzzTime: Long = 500,
    var alloc: TheAllocator = new TheAllocator(),
    var workList: List[() => Unit] = List()) {
  
  val serviceThreads = new ArrayStack[Runnable]
  var pool: PoolType = null;
  @volatile var workers: Option[List[Thread]] = None
  
  def withService[X](
      interval: Long,
      init: X,
      step: (X => Unit),
      inc: (X => X)): Unit = {
    serviceThreads += new Runnable {
      def run() {
        var x = init
        while (sleep()) {
          step(x)
          x = inc(x)
        }
      }
      
      def sleep() = {
        try {
          Thread.sleep(interval)
          true
        } catch {
        case e => false
        }
      }
    }
  }
  
  def everyMs(interval: Long, act: Fuzzer => Unit) {
    withService(interval, this, act, identity: (Fuzzer => Fuzzer))
  }
  
  def countEveryMs(interval: Long, act: Int => Unit) {
    withService(interval, 1, act, (n: Int) => n+1)
  }
  
  def unparkThreads() {
    workers.foreach(_.foreach(LockSupport.unpark))
  }
  
  def doClaims(claims: Int, timeout: Timeout) {
    val deadline = timeout.getDeadline()
    val obj = pool.claim(timeout)
    val excessTime = timeout.getTimeLeft(deadline)
    if (obj == null && excessTime > 0) {
      val unit = timeout.getBaseUnit()
      val msg = "claimed 'null' with " + excessTime + " " + unit + " left."
      throw new IllegalStateException(msg)
    }
    try {
      if (claims > 0) {
        doClaims(claims - 1, timeout)
      }
    } finally {
      if (obj != null) {
        obj.release()
      }
    }
  }
  
  def claimRelease(count: Int, timeout: Timeout) {
    val f = () => {
      doClaims(count, timeout)
    }
    workList = f :: workList
  }
  
  def allocationFailureRate(rate: Double) {
    alloc.setAllocationFailureRate(rate)
  }
  
  def deallocationFailureRate(rate: Double) {
    alloc.setDeallocationFailureRate(rate)
  }
  
  def setTargetSize(size: Int) {
    pool.setTargetSize(size)
  }
  
  def fuzz() {
    config.setSize(poolSize)
    config.setAllocator(alloc)
    pool = poolFactory(config)
    
    val relativeThreadCount = (poolSize * relativeThreadFactor)
    val threadCount = math.max(absoluteThreadCount, relativeThreadCount).toInt
    
    val startLatch = new CountDownLatch(1)
    val workers = buildWorkers(startLatch, threadCount)
    workers.foreach(_.start())
    startLatch.countDown()
    Thread.sleep(fuzzTime)
    workers.foreach(_.interrupt())
    workers.foreach(_.join())
    
    if (!pool.shutdown().await(new Timeout(10, TimeUnit.SECONDS))) {
      throw new IllegalStateException("Pool did not shut down!")
    }
    
    alloc.validateAllocations()
    
    workList = List()
    alloc.reset()
  }
  
  def buildWorkers(latch: CountDownLatch, threadCount: Int): List[Thread] = {
    val workers = workList.map(worker(latch, _))
    workers
  }
  
  val workerCounter = new AtomicInteger
  def worker(latch: CountDownLatch, work: () => Unit): Thread = {
    val script: Runnable = new Runnable {
      def run() {
        latch.await()
        try {
          doWork()
        } catch {
          case e: InterruptedException => ()
        }
      }
      
      def doWork() {
        while (!Thread.currentThread().isInterrupted()) {
          for (_ <- 1 to 100) {
            try {
              work()
            } catch {
              case e: PoolException => if (e.getCause() != TheAllocator.expectableException)
                errorT(e)
            }
          }
        }
      }
    }
    new Thread(script, "Worker-" + workerCounter.incrementAndGet())
  }
}
